File: 065/cpp/main.cpp
enum Token {
    SIGN,
    DOT,
    SCIENCE,
    DIGIT,
    OTHER,
    END,
};

/*
 * INTERGER_WITHOUT_SCIENCE = [+|-]digit+
 * NUMBER_WITHOUT_SCIENCE   = [+|-]digit*[.digit*]
 * NUMBER                   = NUMBER_WITHOUT_SCIENCE[(e|E)INTERGER_WITHOUT_SCIENCE]
 */
class Solution {
public:
    Token getToken(string s, int &i) {
        if (i >= s.size()) {
            return Token::END;
        }
        char c = s[i];
        if (c == '+' || c == '-') {
            return Token::SIGN;
        } else if (c == '.') {
            return Token::DOT;
        } else if (c == 'e' || c == 'E') {
            return Token::SCIENCE;
        } else if (isdigit(c)) {
            return Token::DIGIT;
        } else {
            return Token::OTHER;
        }
    }

    bool consume(string s, int &i, Token token) {
        Token token0 = getToken(s, i);
        if (token0 == token) {
            i++;
            return true;
        }
        return false;
    }

    bool isIntegerWithoutScience(string s, int &i) {
        int n = 0;
        consume(s, i, Token::SIGN);
        while (true) {
            if (!consume(s, i, Token::DIGIT)) {
                break;
            }
            n++;
        }
        return n > 0;
    }

    bool isNumberWithoutScience(string s, int &i) {
        int n = 0;
        consume(s, i, Token::SIGN);
        while (true) {
            if (!consume(s, i, Token::DIGIT)) {
                break;
            }
            n++;
        }
        consume(s, i, Token::DOT);
        while (true) {
            if (!consume(s, i, Token::DIGIT)) {
                break;
            }
            n++;
        }
        return n > 0;
    }

    bool isNumber(string s) {
        int i = 0;
        if (!isNumberWithoutScience(s, i)) {
            return false;
        }
        if (consume(s, i, Token::SCIENCE)) {
            if (!isIntegerWithoutScience(s, i)) {
                return false;
            }
        }
        return consume(s, i, Token::END);
    }
};

File: 065/rust/main.rs
#[derive(Debug,PartialEq)]
enum Token {
    SIGN,
    DOT,
    SCIENCE,
    DIGIT,
    OTHER,
    EOF,
}

impl Solution {
    pub fn get_token(s: &[u8], i: &mut usize) -> Token {
        if *i >= s.len() {
            return Token::EOF;
        }
        let c = s[*i] as char;
        if c == '+' || c == '-' {
            return Token::SIGN;
        } else if c == '.' {
            return Token::DOT;
        } else if c == 'e' || c == 'E' {
            return Token::SCIENCE;
        } else if c.is_digit(10) {
            return Token::DIGIT;
        }
        return Token::OTHER;
    }

    pub fn consume(s: &[u8], i: &mut usize, token: Token) -> bool {
        let token0 = Solution::get_token(s, i);
        if token0 == token {
            *i += 1;
            return true;
        }
        return false;
    }

    pub fn is_integer_without_science(s: &[u8], i: &mut usize) -> bool {
        let mut n = 0;
        Solution::consume(s, i, Token::SIGN);
        loop {
            if !Solution::consume(s, i, Token::DIGIT) {
                break;
            }
            n += 1;
        }
        return n > 0;
    }

    pub fn is_number_without_science(s: &[u8], i: &mut usize) -> bool {
        let mut n = 0;
        Solution::consume(s, i, Token::SIGN);
        loop {
            if !Solution::consume(s, i, Token::DIGIT) {
                break;
            }
            n += 1;
        }
        Solution::consume(s, i, Token::DOT);
        loop {
            if !Solution::consume(s, i, Token::DIGIT) {
                break;
            }
            n += 1;
        }
        return n > 0;
    }

    pub fn is_number(s: String) -> bool {
        let s = s.as_bytes();
        let mut i = 0;
        if !Solution::is_number_without_science(s, &mut i) {
            return false;
        }
        if Solution::consume(s, &mut i, Token::SCIENCE) {
            if !Solution::is_integer_without_science(s, &mut i) {
                return false;
            }
        }
        return Solution::consume(s, &mut i, Token::EOF);
    }
}
